Hw5 is done

Ltaudato1 opened the pull request.

Ltaudato1 committed:
hw5 is done

Ltaudato1 committed:
hw6 is done

anatolik2509 commented on scrapper/src/main/java/edu/java/domain/ChatsJdbcDao.java:
```diff
@@ -0,0 +1,35 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public ChatResponse add(RegisterChatRequest chat) {
```
Прикол трехзвенной архитектуры в том, чтобы изолировать по слоям зоны ответственности. В идеале это выглядит так:
1. Слой контроллеров оперирует запросами ответами, достает из них нужные данные и передаёт их в слой сервисов
2. Слой сервисов отвечает чисто за бизнес-логику и по идее не должен ничего знать ни про запросы-ответы из http, ни про то, каким образом всё это в бд сохраняется
3. Слой dao отвечает за сохранение данных в бд

Это всё позволяет делать компоненты приложения слабосвязанными, значит если вдруг потребуется переделать один из слоев, то мы легко это сможем его переписать, не задевая другие слои.
Ну в общем это к тому, что про объекты RegisterChatRequest и ChatResponse должен знать только слой контроллеров, слой dao про них ничего знать не должен

anatolik2509 commented on scrapper/src/main/java/edu/java/service/jdbc/JdbcLinkService.java:
```diff
@@ -0,0 +1,47 @@
+        List<LinkResponse> links = linksJdbcDao.findAll();
+
+        return linksJdbcDao.findAll().stream()
+            .filter(link -> ChronoUnit.MINUTES.between(link.getLastUpdateTime(), OffsetDateTime.now()) > CHECK_INTERVAL)
```
По идее задания фильтрацию надо было на уровне dao сделать в sql запросе(

anatolik2509 commented on scrapper/src/main/java/edu/java/service/jdbc/JdbcLinkUpdateService.java:
```diff
@@ -0,0 +1,74 @@
+    public int update(Long linkId) {
+        OffsetDateTime currentTime = OffsetDateTime.now();
+
+        String sql = "UPDATE links SET last_update_time = ? WHERE id = ?";
```
Опять таки это должно быть на dao слое, а этот класс на слое сервисов

Ltaudato1 committed:
hw7 is done

Ltaudato1 committed:
hw8 is done

Ltaudato1 committed:
hw9

Ltaudato1 committed:
hw10

Ltaudato1 committed:
updated workflow

Ltaudato1 committed:
added jars

Ltaudato1 committed:
ACTUALLY added jars

Ltaudato1 committed:
updated docker-files

Ltaudato1 committed:
temporary fix

Ltaudato1 committed:
temporary fix v2

почему-то именно на гите докер билд вообще не проходит, хотя локально всё работает вообще отлично

Ltaudato1 committed:
refactor hw5

Ltaudato1 committed:
fix checkstyle

Ltaudato1 committed:
refactor hw7

Vladislav-Kisliy commented on bot/src/main/java/edu/java/bot/BotController.java:
```diff
@@ -14,21 +12,16 @@
+                UserMessage.updateMessage(id, linkUpdate.getUrl());
+            }
             return ResponseEntity.ok("Update was processed");
         } catch (Exception e) {
             throw e;
```
затруднительно понять, в чем логика такой обработки ошибок. Может просто не ловить эксепшен?

Vladislav-Kisliy commented on bot/src/main/java/edu/java/bot/user/UserMessage.java:
```diff
@@ -66,6 +67,17 @@ public void noSuchLink(long chatId) {
+        try {
+            bot.execute(sendMessage);
+        } catch (Exception e) {
+            log.error("an error occurred while sending message in chat with ID=" + chatId);
```
не надо использовать в логгерах конкатенацию
https://stackoverflow.com/questions/10555409/logger-slf4j-advantages-of-formatting-with-instead-of-string-concatenation

Vladislav-Kisliy commented on scrapper/src/main/java/edu/java/clients/stackoverflow/StackOverflowClient.java:
```diff
@@ -31,5 +32,11 @@ public StackOverflowUpdate getQuestionUpdates(String questionId, OffsetDateTime
             .get(0);
     }
+
+    public static String getQuestionIdFromLink(String link) {
+        int startIndex = link.indexOf("/questions/") + "/questions/".length();
+        int endIndex = link.indexOf("/", startIndex);
+        return link.substring(startIndex, endIndex);
+    }
```
это же классический утилитный метод, ему точно не место в контроллере. Стоит вынести в отдельный класс + у него явные проблемы с null и пустыми строками

Vladislav-Kisliy reviewed (commented):
надо добавить, несколько мелких правок

Ltaudato1 committed:
Merge pull request #11 from Ltaudato1/hw10

hw10

Ltaudato1 committed:
Merge pull request #10 from Ltaudato1/hw9

hw9

Ltaudato1 committed:
Merge pull request #9 from Ltaudato1/hw8

hw8 is done

Ltaudato1 committed:
Merge pull request #8 from Ltaudato1/hw7

hw7 is done

Ltaudato1 committed:
solving merge conflict pt.1

Ltaudato1 committed:
Merge branch 'hw6' of https://github.com/Ltaudato1/tinkoff_edu_spring into hw6

Ltaudato1 committed:
Merge branch 'hw6' into hw5
